Log progress when benchmarking multiple URLs

felixarntz opened the pull request:
This PR enhances the `benchmark-server-timing` and `benchmark-web-vitals` commands in two ways:
1. In case a file is provided (i.e. most likely multiple URLs are being benchmarked), the commands will now log progress.
2. Errors in the process are now caught, which is particularly relevant for `benchmark-web-vitals`. Puppeteer errors such as navigation timeouts could otherwise lead to the entire process getting stuck.

While logging isn't really needed for benchmarking a single URL, when benchmarking several URLs the lack of it is a poor experience, as you have no idea on the progress. The second point is particularly relevant for that, as without that it could be possible for that progress to even get completely stuck, and you would never find out about it.

westonruter commented on cli/lib/cli/args.mjs:
```diff
@@ -60,3 +60,7 @@ export async function* getURLs( opt ) {
 }
+
+export function logURLProgress( opt ) {
+	return ! opt.url && !! opt.file;
```
It could be that there is only a single URL in the file. Would that be a problem?

felixarntz replied:
I considered that, but it's not really a big deal. Logging too much isn't an issue, but I think it's reasonable to expect that when a file is provided you want to benchmark multiple URLs. Otherwise you could just provide the URL.

Checking whether there's actually one URL in the file would be overly complicated and involve changing how the file contents are parsed, which right now relies on `yield` for more efficient handling. So I think it's not worth going into that.

westonruter replied:
OK, sounds good. Still, it could be useful even when checking a single `--url`. In that case, what about instead of conditionally logging based on whether `--file` is provided, that there be a new `--verbose` option?

felixarntz replied:
I would prefer not to introduce that for now, unless needed. Why would the existing logging be useful for a single URL? I agree logging would be useful if there are multiple runs, and that shouldn't be coupled to how many URLs are being benchmarked. But the information of "Benchmarking {url}" doesn't provide any value IMO if it's only a single URL.

westonruter replied:
Right, this only makes sense when showing progress for multiple runs of a single URL.

westonruter commented on cli/lib/cli/args.mjs:
```diff
@@ -60,3 +60,7 @@ export async function* getURLs( opt ) {
 	}
 }
+
+export function logURLProgress( opt ) {
```
Maybe this should be renamed to reflect its boolean nature, like `shouldLogProgress`. Otherwise the name sounds like _it_ is responsible for the logging.

westonruter commented on cli/commands/benchmark-server-timing.mjs:
```diff
@@ -83,16 +83,28 @@ export async function handler( opt ) {
-			}
-		);
+		if ( logProgress ) {
+			log( `Benchmarking URL ${ url }...` );
```
Instead of logging out to STDOUT, can this instead go to STDERR? In that way, if you pipe the output into a file, it will by default not include the logging information. This can be achieved by using `console.warn()` instead of `console.log()`. Or it could use `process.stderr.write()`.

felixarntz replied:
What would be the benefit of that? You'd still end up with the "Success." in the file. Or even if we also moved that to STDOUT, the actual errors would end up in the file.

There is probably room to improve this, but if we want to do so, I think we should enhance the existing `log` and `formats` functions available rather than hard-coding in a specific command.

westonruter replied:
The benefit is that the actual output wouldn't get intermixed with the logging. For example, you could do:

```bash
npm run --silent research benchmark-server-timing -- --file=urls.txt --output=csv > results.csv
```

When sending logging information to STDERR, then the CSV file will be pristine and the logging information will only be displayed in the console. Same goes for "Success" below.

And STDERR is not just for errors. It's also for diagnostics. cf. https://stackoverflow.com/a/4919110/93579

felixarntz replied:
Makes sense. So I guess we could use either `console.warn()` or `process.stderr.write()`.

westonruter commented on cli/commands/benchmark-web-vitals.mjs:
```diff
@@ -161,13 +161,28 @@ export async function handler( opt ) {
-			params
-		);
+		if ( logProgress ) {
+			log( `Benchmarking URL ${ url }...` );
```
Using `process.stderr.write` won't automatically add a newline, so if doing this:

```suggestion
			process.stderr.write( `Benchmarking URL ${ url }...` );
```

Then down below the success/error can be displayed on the same line.

felixarntz replied:
It looks like this would also be the case for `process.stdout.write`, this is related to the other comments above. But I definitely like the idea of keeping this in one line.

I'm not sure what is the tradeoffs between using `console.*` functions vs `process.stdout/stderr` in this context. We should probably discuss that and then decide how to proceed by updating the `log` and `formats` functions accordingly.

westonruter commented on cli/commands/benchmark-web-vitals.mjs:
```diff
@@ -161,13 +161,28 @@ export async function handler( opt ) {
+			);
+			results.push( [ url, completeRequests, metrics ] );
+			if ( logProgress ) {
+				log( formats.success( 'Success.' ) );
```
```suggestion
				process.stderr.write( formats.success( `Success\n` ) );
```

westonruter commented on cli/commands/benchmark-web-vitals.mjs:
```diff
@@ -161,13 +161,28 @@ export async function handler( opt ) {
+				log( formats.success( 'Success.' ) );
+			}
+		} catch ( err ) {
+			log( formats.error( `Error: ${ err.message }.` ) );
```
```suggestion
			process.stderr.write( formats.error( `Error: ${ err.message }.\n` ) );
```

westonruter commented on cli/commands/benchmark-web-vitals.mjs:
```diff
@@ -161,13 +161,28 @@ export async function handler( opt ) {
-		results.push( [ url, completeRequests, metrics ] );
+		// Catch Puppeteer errors to prevent the process from getting stuck.
+		try {
+			const { completeRequests, metrics } = await benchmarkURL(
```
What about progress for benchmarking an individual URL? If I have 100 requests for each URL, I'd love to be able to see progress for how far along it is when analyzing each URL.

felixarntz replied:
Makes sense to add this. I would suggest we only do so if more than 1 request is being made, as otherwise there wouldn't be any value.

felixarntz commented:
@westonruter I've only addressed one of your feedback points so far, since I think the others deserve a bit more conversation on how to best proceed. For example regarding https://github.com/GoogleChromeLabs/wpp-research/pull/79#discussion_r1374891590 and https://github.com/GoogleChromeLabs/wpp-research/pull/79#discussion_r1373933060.

Let's figure that out here. Once we've made a decision, we can update / expand the log functions in the helper file accordingly, and then the commands' use of them.

felixarntz commented:
@westonruter I've thought about the logging and output separation a bit more based on your suggestions. How about we implement the following functions:

```js
// Responsible for the actual command output.
export const output = ( text ) => {
	process.stdout.write( `${ text }\n` );
};

// Responsible for (debug) logging.
export const log = ( text ) => {
	process.stderr.write( `${ text }\n` );
};

// Responsible for (debug) logging without "terminating" the line.
export const logPartial = ( text ) => {
	process.stderr.write( text );
};
```

We could then update the usage to rely on the relevant functions. Most importantly, we'd need to change the actual output functions of the command to use `output()`. Let me know what you think - also happy to discuss alternative function names.

westonruter commented:
@felixarntz Those functions look good to me.

felixarntz commented:
@westonruter I have added the new more granular logging functions, updated the commands to use them correctly, and I've also added support for logging multiple `benchmark-web-vitals` iterations as requested in https://github.com/GoogleChromeLabs/wpp-research/pull/79#discussion_r1373935911. Can you please take another look?

westonruter commented on cli/lib/cli/logger.mjs:
```diff
@@ -23,7 +23,20 @@ import chalk from 'chalk';
 import { stringify as formatCsv } from 'csv-stringify/sync'; // eslint-disable-line import/no-unresolved
 
-export const log = console.log; // eslint-disable-line no-console
+// Responsible for the actual command output.
+export const output = ( text ) => {
+	process.stdout.write( `${ text }\n` );
+};
+
+// Responsible for (debug) logging.
+export const log = ( text ) => {
+	process.stderr.write( `${ text }\n` );
+};
+
+// Responsible for (debug) logging without "terminating" the line.
+export const logPartial = ( text ) => {
+	process.stderr.write( text );
+};
```
Nit: The comments could be formatted as jsdoc so that they show up when inspecting.

westonruter reviewed (approved):
Love the logging of the iterations!
